handle dependencies in Envoy filters

ramaraochavali opened the pull request:
This PR handles dependencies in envoy filters - for example if the subfilter is also being inserted as envoy filter - In this example we want to insert rate limiting filter after ext authz and ext authz is also inserted as another envoy filter.

```
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  annotations:
  name: rls-rl-filter
  namespace: rls-ns
spec:
  configPatches:
  - applyTo: HTTP_FILTER
    match:
      context: SIDECAR_INBOUND
      listener:
        filterChain:
          transport_protocol: "tls"
          filter:
            name: "envoy.filters.network.http_connection_manager"
            sub_filter:
               name: "envoy.filters.http.ext_authz"
    patch:
      operation: INSERT_AFTER
      value:
        name: envoy.filters.http.ratelimit
        typed_config:
          '@type': type.googleapis.com/envoy.extensions.filters.http.ratelimit.v3.RateLimit
      
```
[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure


Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[X] Does not have any changes that may affect Istio users.

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | 36dd0a64cbdff5a35885b51a512946c8eb272ecf | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/31891/release-notes_istio/1378614722156302336) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test release-notes_istio

mandarjog commented on pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch.go:
```diff
@@ -409,43 +409,20 @@ func doHTTPFilterListOperation(patchContext networking.EnvoyFilter_PatchContext,
 		doHTTPFilterOperation(patchContext, patches, listener, fc, filter, httpFilter, &httpFiltersRemoved)
 	}
 	for _, cp := range patches[networking.EnvoyFilter_HTTP_FILTER] {
+		// We should process INSERT_AFTER patches at the end so that we can apply them after other patches.
```
Same rationale applies to insert-before, or anything else.
The real feature at hand is *dependency* between envoy filters.
And the solution we are looking for is to apply the dependencies *before* applying modifications.

I don’t think just fixing it for insert-after is Enough.

ramaraochavali replied:
Agreed. Just fixing it for insert-after is enough. But could not think of better way to handle both -  Do you have any suggestions?

mandarjog replied:
The current state is the older resource is applied first everything else being the same. In absence of this change the insert-after patch will fail since the relative anchor does not exist yet. 

Thinking out aloud, if a patch is rejected due to missing anchor, then we look thru the remaining patches that may add that anchor and apply that. If that patch is also relative to another anchor we search and apply that, and so on.

this way we do not change the current filter order unless the current processing is resulting in filter rejection. Changing filter order will be a breaking change.

I think this is a bit complicated, but correct. Wdyt?

ramaraochavali replied:
Yes. I thought about that approach as you mentioned it is a bit complicated. Let me attempt that and see where we go.

kyessenov replied:
Can we just rely on the order of EnvoyFilters? Just name the patches so that dependency comes before use.

ramaraochavali commented:
@mandarjog Changed so that it works for both INSERT_AFTER and INSERT_BEFORE. PTAL. If it looks good, I will add more tests.

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | 0a53c03ee473d73698973a6048647d232aa43f78 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/31891/lint_istio/1379763434320564224) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | 0a53c03ee473d73698973a6048647d232aa43f78 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/31891/lint_istio/1379763434320564224) | `/test lint_istio`
unit-tests_istio | 0a53c03ee473d73698973a6048647d232aa43f78 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/31891/unit-tests_istio/1379763430122065920) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
@mandarjog Updated the code and added tests to verify the dependencies. PTAL.

mandarjog commented on pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch.go:
```diff
@@ -430,31 +429,80 @@ func patchHTTPFilters(patchContext networking.EnvoyFilter_PatchContext,
+			if lp == rp {
+				continue
+			}
+			if hasCyclicDependency(lp, rp) {
```
I don’t think this is complete. We are are not handling a transitive cycle.

ramaraochavali replied:
Yeah. I am afraid of adding another recursion here. But let me add.

ramaraochavali replied:
Fixed the transitive cyclic dependency.

mandarjog commented on pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch.go:
```diff
@@ -430,31 +429,80 @@ func patchHTTPFilters(patchContext networking.EnvoyFilter_PatchContext,
+		}
+		mpatches = append(mpatches, lp)
+	}
+
```
Note: n^2 loop, though expected patches are normally small.

ramaraochavali replied:
yes. That is what I was assuming when I coded this

mandarjog reviewed (commented):
Thanks Rama.
Need more time to review this.

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-test_istio | a68068478cf12f6ab6387dceb62879fc99030340 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/31891/release-test_istio/1381192546478198784) | `/test release-test_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test release-test_istio

mandarjog commented on pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch.go:
```diff
@@ -430,31 +430,67 @@ func patchHTTPFilters(patchContext networking.EnvoyFilter_PatchContext,
+		cycles := hasCyclicDependencies(iapatches)
+		if cycles {
+			IncrementEnvoyFilterMetric(filterKey, HttpFilter, applied)
+			log.Warnf("envoy filter %s has patches that have cyclic dependencies. So filter is skipped", filterKey)
```
What is the filterkey here? Is it enough to identify the problem ?

mandarjog replied:
we still want to apply independent patches?
Adding a cycle should not remove all filters.

mandarjog commented on pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch.go:
```diff
@@ -430,31 +430,67 @@ func patchHTTPFilters(patchContext networking.EnvoyFilter_PatchContext,
+		}
+
+		// Order patches as per the dependencies.
+		for _, lp := range mpatches {
```
Trying to understand why is a recursive function called in a loop.

mandarjog replied:
Dependency map computed in hasCyclic() would be useful here.

mandarjog commented on pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch.go:
```diff
@@ -531,28 +567,107 @@ func patchHTTPFilters(patchContext networking.EnvoyFilter_PatchContext,
+	// dependencies, that means we have a circular dependency.
+	for len(dependencies) != 0 {
+		// Get all filters from the dependencies which have no dependencies
+		independent := sets.NewSet()
```
I think we already have converted partial order based on dependencies to full order here.
if we just append “independent” in a top level array, that array gives us the order in which filters should be applied. Every item in the list is independent of items that come later.

We do not need the 2nd orderDependencies function.

mandarjog replied:
Thinking about it more, the children depend on the parent. So the parent must be evaluated first.

Len(deps)==0 means that no other filter depends on **it**
that does not mean the filter itself is independent. It could depend on something else right?

I think if we can build a tree after we confirm that there are no cycles, then just breadth first traversal is the order we want.

ramaraochavali replied:
Yes. We can not use that map as is as but we can build a tree after there are no cycles and follow breath first travels. Let me try that.

mandarjog commented on pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch.go:
```diff
@@ -531,28 +567,107 @@ func patchHTTPFilters(patchContext networking.EnvoyFilter_PatchContext,
+	for len(dependencies) != 0 {
+		// Get all filters from the dependencies which have no dependencies
+		independent := sets.NewSet()
+		for name, deps := range dependencies {
```
I think the name is confusing here. Either I am not reading it correctly or the name is incorrect.

“””
for filtername, dependents := range dependencies {
 “””
If len(dependents) == 0 it does not make the filter independent.
The cycle detection may still be correct.

ramaraochavali replied:
looks like dependencies is causing confusing. Let me rename that.

ramaraochavali commented:
@mandarjog addressed all the comments and cleaned up a bit. PTAL. Only thing remaining is when cycle is detected it rejects entire patch - instead we should try to remove cycles and apply others , but I need to think about it more as applying rest makes sense or it might incorrectly behave against what user thinks. So left a TODO for it. PTAL.

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | fc0f0b7dc93850762db75b5deeb51e71f8db6cc9 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/31891/lint_istio/1382600103302270976) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | fc0f0b7dc93850762db75b5deeb51e71f8db6cc9 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/31891/lint_istio/1382600103302270976) | `/test lint_istio`
integ-pilot-multicluster-tests_istio | fc0f0b7dc93850762db75b5deeb51e71f8db6cc9 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/31891/integ-pilot-multicluster-tests_istio/1382600099951022080) | `/test integ-pilot-multicluster-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-multicluster-tests_istio | fc0f0b7dc93850762db75b5deeb51e71f8db6cc9 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/31891/integ-pilot-multicluster-tests_istio/1382600099951022080) | `/test integ-pilot-multicluster-tests_istio`
lint_istio | f0f79f84893873526b56ecaf385795277562d7ce | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/31891/lint_istio/1382609053514993664) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-multicluster-tests_istio | 39949ea49963e475f4cc7157ac97d8714d83b2cb | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/31891/integ-pilot-multicluster-tests_istio/1382659625207205888) | `/test integ-pilot-multicluster-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test integ-pilot-multicluster-tests_istio

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-multicluster-tests_istio | 39949ea49963e475f4cc7157ac97d8714d83b2cb | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/31891/integ-pilot-multicluster-tests_istio/1382686551967797248) | `/test integ-pilot-multicluster-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test integ-pilot-multicluster-tests_istio

ramaraochavali commented:
@mandarjog PTAL when you get chance

ramaraochavali commented:
@mandarjog Can you PTAL when you get chance?

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-security-multicluster-tests_istio | 942bcd53e12306250d3c04f393d54c53374e8452 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/31891/integ-security-multicluster-tests_istio/1385179276612997120) | `/test integ-security-multicluster-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test integ-security-multicluster-tests_istio

kyessenov commented:
I'm uneasy to add this much complexity to EnvoyFilter. The patching logic is by design order-dependent (JSON patches don't commute). As long as dependency are specified in the right order, it should be fine without trying to infer the topological ordering? The problem with inference is that it makes the order of application less clear (is it even deterministic?) and the ordering of EnvoyFilters matters even without dependencies.

ramaraochavali commented:
@kyessenov How do you control two teams applying filters? What do you suggest for such cases?

Look at the example I gave 

```
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  annotations:
  name: rls-rl-filter
  namespace: rls-ns
spec:
  configPatches:
  - applyTo: HTTP_FILTER
    match:
      context: SIDECAR_INBOUND
      listener:
        filterChain:
          transport_protocol: "tls"
          filter:
            name: "envoy.filters.network.http_connection_manager"
            sub_filter:
               name: "envoy.filters.http.ext_authz"
    patch:
      operation: INSERT_AFTER
      value:
        name: envoy.filters.http.ratelimit
        typed_config:
          '@type': type.googleapis.com/envoy.extensions.filters.http.ratelimit.v3.RateLimit
```

In this case ext_authz is created by a authorization platform team and ratelimiting created by service team. So service team wants to ensure rate limit is inserted after ext authz - Currently it fails.

kyessenov commented:
@ramaraochavali Using well ordered names for EnvoyFilters would work, no? E.g. let ext_authz be called 001-authz-team and rate limiting be called 002-rate-limit. This way ext-authz is present when processing rate limit filter, and can serve as an anchor for positioning rate limit filter. If you let folks arbitrarily add dependencies, then I'm not sure how you can possibly control dependency loops between teams.

ramaraochavali commented:
> Using well ordered names for EnvoyFilters would work, no? E.g. let ext_authz be called 001-authz-team and rate limiting be called 002-rate-limit. This way ext-authz is present when processing rate limit filter, and can serve as an anchor for positioning rate limit filter. 

I have not tried that - It might work because the Lister will present in alpha order and we are not messing up the order during conversion. Is that what we want to recommend people though @mandarjog What do you think?

> If you let folks arbitrarily add dependencies, then I'm not sure how you can possibly control dependency loops between teams.If you let folks arbitrarily add dependencies, then I'm not sure how you can possibly control dependency loops between teams.
:-) That is what this PR is trying to solve

mandarjog commented:
@kyessenov this PR changes the implementation to match what people assume in the api anyway. It also ensures that loops are detected and discarded. My only hesitation is the complexity of the implementation.

promoting name based ordering as an api, while leaving the current behaviour somewhat undefined isn’t gray either.

If the implementation complexity is not too bad, then this is just a “fix” without changing the api.

ramaraochavali commented:
If the implementation complexity is a worry for us - We could hide behind a feature flag till it is fully tested. If go with the expectation of explicit ordering by name - we need to document it in api so that it is clear.

ramaraochavali commented:
@mandarjog @kyessenov Any thoughts on which route we should take? I am inclined to move forward one way or other with this (my preference would be to add this fix with flag as it is complicated and might be risky and that is how it is documented) - we are getting lot of pressure from service owners as they think it is a Istio bug. But if we all think it is not worth adding that complexity and should be taken care by naming - we should document it so that there is no confusion.

mandarjog commented:
If we are to use  `001-names`, we need to have explicit `sort()` and not rely on the k8s library.

ramaraochavali commented:
Yes. Agree @mandarjog

kyessenov commented:
The way I see it, we need clarity on the order of application of EnvoyFilter regardless of this PR. The dependencies are necessary for any REPLACE operation, since it's important to know which one comes last, and it's not possible to express it with the API I think. If the simple alphanum order works, then we should just rely on it. The concern is how this interacts with namespaces (should global EnvoyFilters come first, should they all be sorted?). This becomes an issue if, for example, the admin wants to make sure their EnvoyFilter comes last.

ramaraochavali commented:
@mandarjog WDYT?

mandarjog commented:
@smawson @nrjpoddar @louiscryan 
the changes we are considering enable EnvoyFilter patches to be applied in an order that a user expects.

Rama has described scenario. Ext_authz is inserted as an EnvoyFilter for everyone. A few teams want to insert a rate limit filter *before* ext authz.

how do we take into account dependencies.

Options
1. Process the dependency first, and then the dependent patch. This pr does that. It adds to the complexity of the code though.
2. User rc.d style sort by patch resource name. 
3. Add a priority field which will be used to order patches.

2 and 3 are api changes. (1) is fixing implementation to match the implied api. Views?

mandarjog commented:
With some fresh perspective I think we should abandon the dependency calculation due to complexity risk.

That leaves us with patch name sorting or adding processing priority order. We will have to explain it in the api doc anyway, so might as well add priority order.

ramaraochavali commented:
> with patch name sorting or adding processing priority order We will have to explain it in the api doc anyway, so might as well add priority order.

Do you mean add an api for priority order? or explain the order would be based on patch name sorts?

ramaraochavali commented:
https://github.com/istio/istio/pull/32581 - to sort filters by name

ramaraochavali commented:
Closing this as we want to move towards name based approach.
